tests/func/api/test_show.py

```diff
@@ -0,0 +1,404 @@
+    dvc.run(name="prepare", cmd="echo preparing data")
+    scm.add(["dvc.yaml", "dvc.lock"])
+    scm.commit("prepare data")
+    sub_dir = tmp_dir / "eval"
+    sub_dir.mkdir()
+    tmp_dir.gen(
+        "tmp_train_val_metrics.json",
+        json.dumps(FIXTURE_METRICS["train_val_metrics"][0]),
+    )
+    run_copy_metrics(
+        "tmp_train_val_metrics.json",
+        "eval/train_val_metrics.json",
+        name="train",
+        metrics_no_cache=["eval/train_val_metrics.json"],
+    )
+    (tmp_dir / "tmp_train_val_metrics.json").unlink()
+
+    scm.add(["dvc.yaml", "dvc.lock", "eval/train_val_metrics.json"])
+    scm.commit("train model")
+
+    tmp_dir.gen("tmp_test_metrics.json", json.dumps(FIXTURE_METRICS["test_metrics"][0]))
+    run_copy_metrics(
+        "tmp_test_metrics.json",
+        "eval/test_metrics.json",
+        name="test",
+        metrics_no_cache=["eval/test_metrics.json"],
+    )
+    (tmp_dir / "tmp_test_metrics.json").unlink()
+
+    scm.add(["dvc.yaml", "dvc.lock", "eval/test_metrics.json"])
+    scm.commit("test model")
+
+    with tmp_dir.branch("better-model", new=True):
+        tmp_dir.gen(
+            "tmp_train_val_metrics.json",
+            json.dumps(FIXTURE_METRICS["train_val_metrics"][1]),
+        )
+        run_copy_metrics(
+            "tmp_train_val_metrics.json",
+            "eval/train_val_metrics.json",
+            name="train",
+            metrics_no_cache=["eval/train_val_metrics.json"],
+        )
+        (tmp_dir / "tmp_train_val_metrics.json").unlink()
+
+        scm.add(["dvc.yaml", "dvc.lock", "eval_train_val_metrics.json"])
+        scm.commit("train better model")
+
+        tmp_dir.gen(
+            "tmp_test_metrics.json", json.dumps(FIXTURE_METRICS["test_metrics"][1])
+        )
+        run_copy_metrics(
+            "tmp_test_metrics.json",
+            "eval/test_metrics.json",
+            name="test",
+            metrics_no_cache=["eval/test_metrics.json"],
+        )
+        (tmp_dir / "tmp_test_metrics.json").unlink()
+
+        scm.add(["dvc.yaml", "dvc.lock", "eval/test_metrics.json"])
+        scm.commit("test better model")
+
+    scm.checkout("master")
+
+
+def test_metrics_show_no_metrics_found(tmp_dir, dvc):
+    # Empty repo
+    with pytest.raises(DvcException, match="No metrics found"):
+        api.metrics_show()
+
+    # dvc.yaml but no metrics
+    dvc.stage.add(name="echo", cmd="echo foo")
+    with pytest.raises(DvcException, match="No metrics found"):
+        api.metrics_show()
+
+
+def test_metrics_show_rev_without_metrics(metrics_repo):
+    with pytest.raises(DvcException, match="No metrics found"):
+        api.metrics_show(rev="HEAD~2")
+
+
+def test_metrics_show_rev_with_metrics(metrics_repo):
+    assert api.metrics_show(rev="HEAD~1") == FIXTURE_METRICS["train_val_metrics"][0]
+    assert api.metrics_show(rev="HEAD") == {
+        "eval/train_val_metrics.json:avg_prec": FIXTURE_METRICS["train_val_metrics"][0][
```

Comment:
Windows tests are failing. 
You need to define the paths using `os.path.join`, probably worth to also do it outside to reuse it across tests

Reply:
I think the issue wasn't necessarily not using `os.path.join`—though I did change the code to use this as well—but rather that the dictionary returned by `metrics_show` on Windows had keys like `"eval//train_val_metrics.json"` (instead of `"eval/train_val_metrics.json"`). So the main change I made to address this was to use `{os.sep}` in an f-string when defining the expected dictionary in the `assert` statements.

The Windows tests are now passing.

Reply:
Sorry for my lack of clarity. I was referring to using it like this:

https://github.com/iterative/dvc/pull/9042/commits/0ddbba9782a5db05baf3d31d9e7ee33c577f4336

It was minor anyhow after test were fixed so I just quickly added it myself

Reply:
Nice, thank you. Wasn't actually aware you could return stuff in a `@pytest.fixture` like this. So I learned something new and agree this is much cleaner. Thanks for the extra help on this! 😄